Archives/VolFile.cpp

```diff
@@ -389,103 +382,102 @@ namespace Archives
-	// Note: tagText is a 4 byte field
-	int VolFile::ReadTag(int offset, const char *tagText)
+	// invalid then an error is thrown.
+	uint32_t VolFile::ReadTag(const std::array<char, 4>& tagName)
 	{
-		char *addr = (char*)m_BaseOfFile + offset;
+		// Tags are 4 chars in length and do not include a null terminator
+		std::array<char, 4> tagFromFile;
+		archiveFileReader->Read(tagFromFile.data(), tagFromFile.size());
 
-		if (*(int*)addr != *(int*)tagText) { 	   // Make sure the tag matches
-			return -1;
+		if (tagFromFile != tagName) {
+			throw std::runtime_error("Tag " + std::string(tagFromFile.data(), tagFromFile.size()) + " does not match requested tag of " + std::string(tagName.data(), tagName.size()));
 		}
 
-		int length = *(int*)(addr + 4);			   // Get the length (and length tag)
-		if ((length & 0x80000000) != 0x80000000) { // Check for the tag (MSB set)
-			return -1;
+		uint32_t length = 0;
+		archiveFileReader->Read(&length, sizeof(length));
+
+		// Check for the tag (MSB set)
+		if ((length & 0x80000000) != 0x80000000) {
+			throw std::runtime_error("Tag " + std::string(tagName.data()) + " contains an invalid length.");
 		}
 
-		return length & 0x7FFFFFFF;				// Return the length (mask out the tag)
+		// Mask out the tag and return the length.
+		return length & 0x7FFFFFFF;					
 	}
 
 	// Reads the header structure of the .vol file and sets up indexing/structure variables
 	// Returns true is the header structure is valid and false otherwise
 	bool VolFile::ReadVolHeader()
 	{
-		int temp;
-
-		// Make sure file is big enough to contain header tag
-		if (m_MappedFileSize < 8) {
-			return false;
-		}
-
-		// Make sure the header is structured properly
-		// -------------------------------------------
-		// -------------------------------------------
+		m_HeaderLength = ReadTag(std::array<char, 4> { 'V', 'O', 'L', ' ' });
 
-		// Check opening header tag ("VOL " tag)
-		// -------------------------------------
-		if ((m_HeaderLength = ReadTag(0, "VOL ")) == -1) {
-			return false;
+		uint32_t volhSize = ReadTag(std::array<char, 4> { 'v', 'o', 'l', 'h' });
+		if (volhSize != 0) {
+			return false; //the size of this section must be 0
 		}
 
-		// Make sure the file is large enough to contain the header
-		if (m_MappedFileSize < m_HeaderLength + 8) {
-			return false;
-		}
+		m_StringTableLength = ReadTag(std::array<char, 4> { 'v', 'o', 'l', 's' });
 
-		// Check next header tag ("volh" tag)
-		// ----------------------------------
-		// Note: the size of this section must be 0
-		if (ReadTag(8, "volh") != 0) {
-			return false;
-		}
-
-		// Check for fileName string table tag ("vols" tag)
-		// ------------------------------------------------
-		if ((m_StringTableLength = ReadTag(16, "vols")) == -1) {
-			return false;
-		}
-
-		m_ActualStringTableLength = *((int*)m_BaseOfFile + 6);	// Store actual length
-		m_StringTable = (char*)m_BaseOfFile + 28;				// Store pointer to table
+		ReadStringTable();
```

Comment:
The string table read should probably happen after the checks on header sizes. Those checks are largely to prevent problems from reading past the end of the file.